lib/cf-socket.c

```diff
@@ -1349,7 +1349,7 @@ static CURLcode cf_tcp_connect(struct Curl_cfilter *cf,
     CURL_TRC_CF(data, cf, "not connected yet");
     return CURLE_OK;
   }
-  else if(rc == CURL_CSELECT_OUT || cf->conn->bits.tcp_fastopen) {
+  else if((rc & CURL_CSELECT_ERR) || rc == CURL_CSELECT_OUT || cf->conn->bits.tcp_fastopen) {
```

Comment:
Should this condition rather be like this:
~~~c
else if((rc & (CURL_CSELECT_ERR|CURL_CSELECT_OUT)) || cf->conn->bits.tcp_fastopen) {
~~~

I take you get *only* an error and yet the socket has connected? It seems like a rather weird system...

Reply:
I just want to reuse the "if(verifyconnect(ctx->sock, &ctx->error))" code snippet. Ignoring cyclomatic complexity, the code is as follows:
```
rc = SOCKET_WRITABLE(ctx->sock, 0);

if(rc == 0) { /* no connection yet */
  CURL_TRC_CF(data, cf, "not connected yet");
  return CURLE_OK;
}
else if(rc == CURL_CSELECT_OUT || cf->conn->bits.tcp_fastopen) {
  if(verifyconnect(ctx->sock, &ctx->error)) {
    /* we are connected with TCP, awesome! */
    ctx->connected_at = Curl_now();
    set_local_ip(cf, data);
    *done = TRUE;
    cf->connected = TRUE;
    CURL_TRC_CF(data, cf, "connected");
    return CURLE_OK;
  }
} else if (rc & CURL_CSELECT_ERR) {
  if(verifyconnect(ctx->sock, &ctx->error)) {
    /* we are connected with TCP, awesome! */
    ctx->connected_at = Curl_now();
    set_local_ip(cf, data);
    *done = TRUE;
    cf->connected = TRUE;
    CURL_TRC_CF(data, cf, "connected");
    return CURLE_OK;
  }
  return CURLE_COULDNT_CONNECT;
}
```

Reply:
> Should this condition rather be like this:
> 
> ```c
> else if((rc & (CURL_CSELECT_ERR|CURL_CSELECT_OUT)) || cf->conn->bits.tcp_fastopen) {
> ```
> 
> I take you get _only_ an error and yet the socket has connected? It seems like a rather weird system...

Actually, the issue isn't a real error. The function SOCKET_WRITABLE uses select to determine writability on iOS, which isn't accurate. When the socket file descriptor exceeds 1024, using select will cause an abort. However, accessing the socket without using the select API works fine. My modification is simply to let the verifyconnect function correct the issues caused by SOCKET_WRITABLE.

Reply:
Then I really cannot accept this change, as it is totally misplaced.

The problem is then that the check needs to be fixed for iOS/AppleOS which we explicitly decide to not use poll() for because they seem to repeatedly break that function.

The question is only: how do you reliably check a socket on these operating systems?